fix: permission checks for editing access keys

donatello opened the pull request:
## Description

With this change, only a user with `UpdateServiceAccountAdminAction` permission is able to edit access keys.

We would like to let a user edit their own access keys, however the feature needs to be re-designed for better security and integration with external systems like AD/LDAP and OpenID.

This change prevents privilege escalation via service accounts.

## Motivation and Context


## How to test this PR?


## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

harshavardhana commented on cmd/iam.go:
```diff
@@ -1848,37 +1848,18 @@ func (sys *IAMSys) IsAllowedServiceAccount(args policy.Args, parentUser string)
+	hasSessionPolicy, isAllowedSP := isAllowedBySessionPolicy(args)
+	if hasSessionPolicy {
+		result := isAllowedSP && (isOwnerDerived || combinedPolicy.IsAllowed(args))
+		fmt.Printf("IsAllowedSTS: hasSessionPolicy: %v, isAllowedSP: %v, result: %v\n", hasSessionPolicy, isAllowedSP, result)
```
```suggestion
```

harshavardhana reviewed (changes requested):
some tests might be needed as well for MinIO as IDP users.

donatello committed:
fix: permission checks for editing access keys

With this change, only a user with `UpdateServiceAccountAdminAction`
permission is able to edit access keys.

We would like to let a user edit their own access keys, however the
feature needs to be re-designed for better security and integration with
external systems like AD/LDAP and OpenID.

This change prevents privilege escalation via service accounts.

klauspost commented on cmd/admin-handlers-users_test.go:
```diff
@@ -896,13 +897,13 @@ func (s *TestSuiteIAM) TestServiceAccountOpsByUser(c *check) {
-	// 4. Check that svc account can restrict the policy, and that the
-	// session policy can be updated.
-	c.assertSvcAccSessionPolicyUpdate(ctx, s, userAdmClient, accessKey, bucket)
+	// // 4. Check that svc account can restrict the policy, and that the
```
Just delete all these. We have git history if we need to find them again.

donatello replied:
Yep.

vadmeste commented on cmd/iam.go:
```diff
@@ -1848,37 +1848,16 @@ func (sys *IAMSys) IsAllowedServiceAccount(args policy.Args, parentUser string)
+	// Now check if we have a sessionPolicy.
+	hasSessionPolicy, isAllowedSP := isAllowedBySessionPolicy(args)
+	if hasSessionPolicy {
+		return isAllowedSP && (isOwnerDerived || combinedPolicy.IsAllowed(args))
```
I may be missing something, but shouldn't this be parentArgs ?

donatello replied:
You're right, fixing!

vadmeste commented on cmd/iam.go:
```diff
@@ -1848,37 +1848,16 @@ func (sys *IAMSys) IsAllowedServiceAccount(args policy.Args, parentUser string)
 
-	return subPolicy.IsAllowed(parentArgs) && (isOwnerDerived || combinedPolicy.IsAllowed(parentArgs))
+	// Sub policy not set. Evaluate only the parent policies.
+	return (isOwnerDerived || combinedPolicy.IsAllowed(args))
```
I may be missing something, but shouldn't this be parentArgs ?

vadmeste reviewed (approved):
LGTM
